File: types/assert_type_test.go
package types

import (
	"bytes"
	"testing"

	"github.com/stretchr/testify/assert"
)

func validateType(t *Type, v Value) {
	assertType(t, v)
}

func assertInvalid(tt *testing.T, t *Type, v Value) {
	assert := assert.New(tt)
	assert.Panics(func() {
		assertType(t, v)
	})
}

func assertAll(tt *testing.T, t *Type, v Value) {
	allTypes := []*Type{
		BoolType,
		NumberType,
		StringType,
		BlobType,
		TypeType,
		ValueType,
	}

	for _, at := range allTypes {
		if at == ValueType || t.Equals(at) {
			validateType(at, v)
		} else {
			assertInvalid(tt, at, v)
		}
	}
}

func TestAssertTypePrimitives(t *testing.T) {
	validateType(BoolType, Bool(true))
	validateType(BoolType, Bool(false))
	validateType(NumberType, Number(42))
	validateType(StringType, NewString("abc"))

	assertInvalid(t, BoolType, Number(1))
	assertInvalid(t, BoolType, NewString("abc"))
	assertInvalid(t, NumberType, Bool(true))
	assertInvalid(t, StringType, Number(42))
}

func TestAssertTypeValue(t *testing.T) {
	validateType(ValueType, Bool(true))
	validateType(ValueType, Number(1))
	validateType(ValueType, NewString("abc"))
	l := NewList(Number(0), Number(1), Number(2), Number(3))
	validateType(ValueType, l)
}

func TestAssertTypeBlob(t *testing.T) {
	blob := NewBlob(bytes.NewBuffer([]byte{0x00, 0x01}))
	assertAll(t, BlobType, blob)
}

func TestAssertTypeList(tt *testing.T) {
	listOfNumberType := MakeListType(NumberType)
	l := NewList(Number(0), Number(1), Number(2), Number(3))
	validateType(listOfNumberType, l)
	assertAll(tt, listOfNumberType, l)
	validateType(MakeListType(ValueType), l)
}

func TestAssertTypeMap(tt *testing.T) {
	mapOfNumberToStringType := MakeMapType(NumberType, StringType)
	m := NewMap(Number(0), NewString("a"), Number(2), NewString("b"))
	validateType(mapOfNumberToStringType, m)
	assertAll(tt, mapOfNumberToStringType, m)
	validateType(MakeMapType(ValueType, ValueType), m)
}

func TestAssertTypeSet(tt *testing.T) {
	setOfNumberType := MakeSetType(NumberType)
	s := NewSet(Number(0), Number(1), Number(2), Number(3))
	validateType(setOfNumberType, s)
	assertAll(tt, setOfNumberType, s)
	validateType(MakeSetType(ValueType), s)
}

func TestAssertTypeType(tt *testing.T) {
	t := MakeSetType(NumberType)
	validateType(TypeType, t)
	assertAll(tt, TypeType, t)
	validateType(ValueType, t)
}

func TestAssertTypeStruct(tt *testing.T) {
	t := MakeStructType("Struct", TypeMap{
		"x": BoolType,
	})

	v := NewStruct("Struct", structData{"x": Bool(true)})
	validateType(t, v)
	assertAll(tt, t, v)
	validateType(ValueType, v)
}

func TestAssertTypeUnion(tt *testing.T) {
	validateType(MakeUnionType(NumberType), Number(42))
	validateType(MakeUnionType(NumberType, StringType), Number(42))
	validateType(MakeUnionType(NumberType, StringType), NewString("hi"))
	validateType(MakeUnionType(NumberType, StringType, BoolType), Number(555))
	validateType(MakeUnionType(NumberType, StringType, BoolType), NewString("hi"))
	validateType(MakeUnionType(NumberType, StringType, BoolType), Bool(true))

	lt := MakeListType(MakeUnionType(NumberType, StringType))
	validateType(lt, NewList(Number(1), NewString("hi"), Number(2), NewString("bye")))

	st := MakeSetType(StringType)
	validateType(MakeUnionType(st, NumberType), Number(42))
	validateType(MakeUnionType(st, NumberType), NewSet(NewString("a"), NewString("b")))

	assertInvalid(tt, MakeUnionType(), Number(42))
	assertInvalid(tt, MakeUnionType(StringType), Number(42))
	assertInvalid(tt, MakeUnionType(StringType, BoolType), Number(42))
	assertInvalid(tt, MakeUnionType(st, StringType), Number(42))
	assertInvalid(tt, MakeUnionType(st, NumberType), NewSet(Number(1), Number(2)))
}

func TestAssertTypeEmptyListUnion(tt *testing.T) {
	lt := MakeListType(MakeUnionType())
	validateType(lt, NewList())
}

func TestAssertTypeEmptyList(tt *testing.T) {
	lt := MakeListType(NumberType)
	validateType(lt, NewList())

	// List<> not a subtype of List<Number>
	assertInvalid(tt, MakeListType(MakeUnionType()), NewList(Number(1)))
}

func TestAssertTypeEmptySet(tt *testing.T) {
	st := MakeSetType(NumberType)
	validateType(st, NewSet())

	// Set<> not a subtype of Set<Number>
	assertInvalid(tt, MakeSetType(MakeUnionType()), NewSet(Number(1)))
}

func TestAssertTypeEmptyMap(tt *testing.T) {
	mt := MakeMapType(NumberType, StringType)
	validateType(mt, NewMap())

	// Map<> not a subtype of Map<Number, Number>
	assertInvalid(tt, MakeMapType(MakeUnionType(), MakeUnionType()), NewMap(Number(1), Number(2)))
}

func TestAssertTypeStructSubtypeByName(tt *testing.T) {
	namedT := MakeStructType("Name", TypeMap{"x": NumberType})
	anonT := MakeStructType("", TypeMap{"x": NumberType})
	namedV := NewStruct("Name", structData{"x": Number(42)})
	name2V := NewStruct("foo", structData{"x": Number(42)})
	anonV := NewStruct("", structData{"x": Number(42)})

	validateType(namedT, namedV)
	assertInvalid(tt, namedT, name2V)
	assertInvalid(tt, namedT, anonV)

	validateType(anonT, namedV)
	validateType(anonT, name2V)
	validateType(anonT, anonV)
}

func TestAssertTypeStructSubtypeExtraFields(tt *testing.T) {
	at := MakeStructType("", TypeMap{})
	bt := MakeStructType("", TypeMap{"x": NumberType})
	ct := MakeStructType("", TypeMap{"x": NumberType, "s": StringType})
	av := NewStruct("", structData{})
	bv := NewStruct("", structData{"x": Number(1)})
	cv := NewStruct("", structData{"x": Number(2), "s": NewString("hi")})

	validateType(at, av)
	assertInvalid(tt, bt, av)
	assertInvalid(tt, ct, av)

	validateType(at, bv)
	validateType(bt, bv)
	assertInvalid(tt, ct, bv)

	validateType(at, cv)
	validateType(bt, cv)
	validateType(ct, cv)
}

func TestAssertTypeStructSubtype(tt *testing.T) {
	c1 := NewStruct("Commit", structData{
		"value":   Number(1),
		"parents": NewSet(),
	})
	t1 := MakeStructType("Commit", TypeMap{
		"value":   NumberType,
		"parents": MakeSetType(MakeUnionType()),
	})
	validateType(t1, c1)

	t11 := MakeStructType("Commit", TypeMap{
		"value":   NumberType,
		"parents": MakeSetType(MakeRefType(NumberType /* placeholder */)),
	})
	t11.Desc.(StructDesc).Fields["parents"].Desc.(CompoundDesc).ElemTypes[0].Desc.(CompoundDesc).ElemTypes[0] = t11
	validateType(t11, c1)

	c2 := NewStruct("Commit", structData{
		"value":   Number(2),
		"parents": NewSet(NewRef(c1)),
	})
	validateType(t11, c2)

	// t3 :=
}

File: js/src/assert-type-test.js
// @flow

import {assert} from 'chai';
import Blob from './blob.js';
import List from './list.js';
import Map from './map.js';
import Set from './set.js';
import {newStruct} from './struct.js';
import {suite, test} from 'mocha';
import assertSubtype from './assert-type.js';
import type {Type} from './type.js';
import {
  blobType,
  boolType,
  listOfValueType,
  makeListType,
  makeMapType,
  makeRefType,
  makeSetType,
  makeStructType,
  makeUnionType,
  mapOfValueType,
  numberType,
  setOfValueType,
  stringType,
  typeType,
  valueType,
} from './type.js';
import {equals} from './compare.js';
import RefValue from './ref-value.js';

suite('validate type', () => {

  function assertInvalid(t: Type, v) {
    assert.throws(() => { assertSubtype(t, v); });
  }

  const allTypes = [
    boolType,
    numberType,
    stringType,
    blobType,
    typeType,
    valueType,
  ];

  function assertAll(t: Type, v) {
    for (const at of allTypes) {
      if (at === valueType || equals(t, at)) {
        assertSubtype(at, v);
      } else {
        assertInvalid(at, v);
      }
    }
  }

  test('primitives', () => {
    assertSubtype(boolType, true);
    assertSubtype(boolType, false);
    assertSubtype(numberType, 42);
    assertSubtype(stringType, 'abc');

    assertInvalid(boolType, 1);
    assertInvalid(boolType, 'abc');
    assertInvalid(numberType, true);
    assertInvalid(stringType, 42);
  });

  test('value', () => {
    assertSubtype(valueType, true);
    assertSubtype(valueType, 1);
    assertSubtype(valueType, 'abc');
    const l = new List([0, 1, 2, 3]);
    assertSubtype(valueType, l);
  });

  test('blob', () => {
    const b = new Blob(new Uint8Array([0, 1, 2, 3, 4, 5, 6, 7]));
    assertAll(blobType, b);
  });

  test('list', () => {
    const listOfNumberType = makeListType(numberType);
    const l = new List([0, 1, 2, 3]);
    assertSubtype(listOfNumberType, l);
    assertAll(listOfNumberType, l);

    assertSubtype(listOfValueType, l);
  });

  test('map', () => {
    const mapOfNumberToStringType = makeMapType(numberType, stringType);
    const m = new Map([[0, 'a'], [2, 'b']]);
    assertSubtype(mapOfNumberToStringType, m);
    assertAll(mapOfNumberToStringType, m);

    assertSubtype(mapOfValueType, m);
  });

  test('set', () => {
    const setOfNumberType = makeSetType(numberType);
    const s = new Set([0, 1, 2, 3]);
    assertSubtype(setOfNumberType, s);
    assertAll(setOfNumberType, s);

    assertSubtype(setOfValueType, s);
  });

  test('type', () => {
    const t = makeSetType(numberType);
    assertSubtype(typeType, t);
    assertAll(typeType, t);

    assertSubtype(valueType, t);
  });

  test('struct', () => {
    const type = makeStructType('Struct', {
      'x': boolType,
    });

    const v = newStruct('Struct', {x: true});
    assertSubtype(type, v);
    assertAll(type, v);

    assertSubtype(valueType, v);
  });

  test('union', () => {
    assertSubtype(makeUnionType([numberType]), 42);
    assertSubtype(makeUnionType([numberType, stringType]), 42);
    assertSubtype(makeUnionType([numberType, stringType]), 'hi');
    assertSubtype(makeUnionType([numberType, stringType, boolType]), 555);
    assertSubtype(makeUnionType([numberType, stringType, boolType]), 'hi');
    assertSubtype(makeUnionType([numberType, stringType, boolType]), true);

    const lt = makeListType(makeUnionType([numberType, stringType]));
    assertSubtype(lt, new List([1, 'hi', 2, 'bye']));

    const st = makeSetType(stringType);
    assertSubtype(makeUnionType([st, numberType]), 42);
    assertSubtype(makeUnionType([st, numberType]), new Set(['a', 'b']));

    assertInvalid(makeUnionType([]), 42);
    assertInvalid(makeUnionType([stringType]), 42);
    assertInvalid(makeUnionType([stringType, boolType]), 42);
    assertInvalid(makeUnionType([st, stringType]), 42);
    assertInvalid(makeUnionType([st, numberType]), new Set([1, 2]));
  });

  test('empty list union', () => {
    const lt = makeListType(makeUnionType([]));
    assertSubtype(lt, new List());
  });

  test('empty list', () => {
    const lt = makeListType(numberType);
    assertSubtype(lt, new List());

    // List<> not a subtype of List<Number>
    assertInvalid(makeListType(makeUnionType([])), new List([1]));
  });

  test('empty set', () => {
    const st = makeSetType(numberType);
    assertSubtype(st, new Set());

    // Set<> not a subtype of Set<Number>
    assertInvalid(makeSetType(makeUnionType([])), new Set([1]));
  });

  test('empty map', () => {
    const mt = makeMapType(numberType, stringType);
    assertSubtype(mt, new Map());

    // Map<> not a subtype of Map<Number, Number>
    assertInvalid(makeMapType(makeUnionType([]), makeUnionType([])), new Map([[1, 2]]));
  });

  test('struct subtype by name', () => {
    const namedT = makeStructType('Name', {x: numberType});
    const anonT = makeStructType('', {x: numberType});
    const namedV = newStruct('Name', {x: 42});
    const name2V = newStruct('foo', {x: 42});
    const anonV = newStruct('', {x: 42});

    assertSubtype(namedT, namedV);
    assertInvalid(namedT, name2V);
    assertInvalid(namedT, anonV);

    assertSubtype(anonT, namedV);
    assertSubtype(anonT, name2V);
    assertSubtype(anonT, anonV);
  });

  test('struct subtype extra fields', () => {
    const at = makeStructType('', {});
    const bt = makeStructType('', {x: numberType});
    const ct = makeStructType('', {x: numberType, s: stringType});
    const av = newStruct('', {});
    const bv = newStruct('', {x: 1});
    const cv = newStruct('', {x: 2, s: 'hi'});

    assertSubtype(at, av);
    assertInvalid(bt, av);
    assertInvalid(ct, av);

    assertSubtype(at, bv);
    assertSubtype(bt, bv);
    assertInvalid(ct, bv);

    assertSubtype(at, cv);
    assertSubtype(bt, cv);
    assertSubtype(ct, cv);
  });

  test('struct subtype', () => {
    const c1 = newStruct('Commit', {
      value: 1,
      parents: new Set(),
    });
    const t1 = makeStructType('Commit', {
      value: numberType,
      parents: makeSetType(makeUnionType([])),
    });
    assertSubtype(t1, c1);

    const t11 = makeStructType('Commit', {
      value: numberType,
      parents: makeSetType(makeRefType(numberType /* placeholder */)),
    });
    t11.desc.fields['parents'].desc.elemTypes[0].desc.elemTypes[0] = t11;
    assertSubtype(t11, c1);

    const c2 = newStruct('Commit', {
      value: 2,
      parents: new Set([new RefValue(c1)]),
    });
    assertSubtype(t11, c2);

    // struct { v: V, p: Set<> } <!
    // struct { v: V, p: Set<Ref<...>> }
    assertInvalid(t1, c2);
  });
});
